### PR TITLE
PrometheusExporter: Project cleanup and middleware fixes

### PR DESCRIPTION
**Pull request by CodeBlanch**
## Changes

Sorry, this one got a little big 😰 

* Cleaned up the project so it more closely follows the patterns used in zipkin, jaeger, otlp, etc.
* Middleware didn't seem to work at all. It works now! Had to re-do the settings to separate the listener host + port and the path. Removed the string/UTF8 byte[] allocation when using the middleware.
* Use async all the way throughout the metric collection serialization. Should improve perf but I haven't measure it yet.
* HttpListener was always started, never stopped. It is now started on .NET Framework by default and NOT started on .NET/Core (assuming most people will use the middleware on .NET/Core). Moved http listener into the exporter itself so it can be stopped on shutdown.
* Seemed to be a race when using middleware and/or http listener in that multiple concurrent requests would each call collect which would change the mertrics held on the exporter. I added a semaphore to prevent this. Concurrent requests will now error out instantly with an HTTP 429 status code. Looking for feedback on this 🤔 

## Future PRs

Planning to do some follow-ups to this:

* Benchmarks
* Unit tests

## TODOs

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by reyang** [comment]
> * Seemed to be a race when using middleware and/or http listener in that multiple concurrent requests would each call collect which would change the mertrics held on the exporter. I added a semaphore to prevent this. Concurrent requests will now error out instantly with an HTTP 429 status code. Looking for feedback on this 🤔

There are couple things I've been thinking while refactoring the MetricReader and MetricExporter:
1. Does PrometheusExporter support sessions? E.g. two scrapers pulling metrics independently without causing side effect to each other? I think this _might_ be achieved by either having two separate exporter instances (and each scraper will only pull from the designated endpoint), or having some sticky session information (e.g. Cookie?).
2. Should MetricReader.Collect be reentrant, and what does that mean? I think YES, MetricReader.Collect can be called simultaneously, and at any given time, only one `Collect` call should do the real work, while all other calls should simply wait and share the outcome. For example:
  * Thread 1 called Collect
  * Thread 2 called Collect - will wait since Thread 1 Collect is running
  * Thread 3 called Collect - will wait since Thread 1 Collect is running
  * Thread 1 Collect finished - Thread 1 Collect will return control to the caller, Thread 2/3 unblocked, now Thread 3 is going to do the job, while Thread 2 waits for 3 to complete so it can share the outcome
  * Thread 3 Collect unblocked
  * Thread 4 called Collect - will wait since Thread 3 Collect is running, so it cannot "catch the train"
  * Thread 3 Collect returned - now Thread 2 Collect is unblocked, and can share the result
  * Thread 2 Collect unblocked - by "stealing the result" from Thread 3
  * Thread 2 Collect returned
  * Thread 4 Collect unblocked - now doing the actual work because it is the only thread asking for Collect at this moment
  * Thread 4 Collect returned

**Comment by CodeBlanch** [comment]
@reyang How do you feel about merging this as-is (logically) and then doing further work to support concurrent pull? IMO what we have today is unsafe/dangerous/undefined so at least this gives us some safety until we can support it fully.

**Review by reyang — APPROVED** [review]

**Comment by reyang** [comment]
> @reyang How do you feel about merging this as-is (logically) and then doing further work to support concurrent pull? IMO what we have today is unsafe/dangerous/undefined so at least this gives us some safety until we can support it fully.

Agree 💯, I've approved the PR.

File: src/OpenTelemetry.Exporter.Prometheus/README.md [hunk 1]
```diff
@@ -7,17 +7,88 @@
...
+    Call the `AddPrometheusExporter` `MeterProviderBuilder` extension to
+    register the Prometheus exporter.
+
+    ```csharp
+    using var meterProvider = Sdk.CreateMeterProviderBuilder()
+        .AddPrometheusExporter()
+        .Build();
+    ```
+
+### Step 3: Register ASP.NET Core Prometheus Scraping Middleware
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Prometheus/README.md:40]: Isn't this optional?

Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Prometheus/README.md]: This PR changes the behavior a bit.

* .NET Framework: It works as it did before. HttpListener is automatically started.

* .NET/Core: HttpListener is NOT automatically started. Users must configure middleware or explicitly set the listener to start.

The reason being, HttpListener goes outside of ASP.NET Core: https://docs.microsoft.com/en-us/dotnet/api/system.net.httplistener?view=net-5.0#httpsys What I'm worried about is customers relying on Kestrel pipeline for filtering, SSL, origin detection, pathBase, whatever, and then the HttpListener circumventing that.

I did update the README to try and make this more clear.

File: examples/Console/TestPrometheusExporter.cs [hunk 2]
```diff
@@ -49,7 +49,11 @@ internal static object Run(int port, int totalDurationInMins)
             */
             using var meterProvider = Sdk.CreateMeterProviderBuilder()
                 .AddSource("TestMeter")
-                .AddPrometheusExporter(opt => opt.Url = $"http://localhost:{port}/metrics/")
+                .AddPrometheusExporter(opt =>
+                {
+                    opt.StartHttpListener = true;
+                    opt.HttpListenerPrefixes = new string[] { $"http://*:{port}/" };
```
Comment (cijothomas) [inline examples/Console/TestPrometheusExporter.cs:55]: nit: there is a message printed to console showing the path where metrics is available, which still has /metrics in it..

**Review by cijothomas — APPROVED** [review]
LGTM.
Most issues from https://github.com/open-telemetry/opentelemetry-dotnet/issues/2400 is covered here!

**Pull request merged**